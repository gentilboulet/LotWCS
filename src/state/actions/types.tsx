import { IInitialStateAction } from './initial';
import { IHeaderAction } from './header';
import { IChiAction } from './chi';
import { ISkillAction } from './skills';
import { ILoresheetAction } from './loresheets';
import { IHistoryAction } from './history';
import { IKungFuAction } from './kungfus';
import { IVirtueAction } from './virtues';

export type IAction =
  IInitialStateAction | IHeaderAction | ISkillAction | IHistoryAction |
  ILoresheetAction | IChiAction | IKungFuAction | IVirtueAction;
