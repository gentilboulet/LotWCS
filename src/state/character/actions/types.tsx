import { ActionType } from "typesafe-actions";

import * as chi from "./chi";
import * as header from "./header";
import * as kungfu from "./kungfu";
import * as loresheets from "./loresheets";
import * as skills from "./skills";
import * as virtues from "./virtues";

export type ICharacterAction =
  | ActionType<typeof chi>
  | ActionType<typeof header>
  | ActionType<typeof kungfu>
  | ActionType<typeof loresheets>
  | ActionType<typeof skills>
  | ActionType<typeof virtues>;
