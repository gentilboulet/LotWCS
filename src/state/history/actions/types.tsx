import { ActionType } from "typesafe-actions";

import * as history from "./history";

export type IHistoryAction = ActionType<typeof history>;
