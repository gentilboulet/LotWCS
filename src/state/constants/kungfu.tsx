export const KUNGFU_OPEN_STYLE = 'KUNGFU_OPEN_STYLE';
export type KUNGFU_OPEN_STYLE = typeof KUNGFU_OPEN_STYLE;

export const KUNGFU_CUSTOM_NAME_FOR_STYLE = 'KUNGFU_CUSTOM_NAME_FOR_STYLE';
export type KUNGFU_CUSTOM_NAME_FOR_STYLE = typeof KUNGFU_CUSTOM_NAME_FOR_STYLE;

export const KUNGFU_CUSTOM_NAME_FOR_TECHNIQUE = 'KUNGFU_CUSTOM_NAME_FOR_TECHNIQUE';
export type KUNGFU_CUSTOM_NAME_FOR_TECHNIQUE = typeof KUNGFU_CUSTOM_NAME_FOR_TECHNIQUE;

export const KUNGFU_BUY_TECHNIQUE = 'KUNGFU_BUY_TECHNIQUE';
export type KUNGFU_BUY_TECHNIQUE = typeof KUNGFU_BUY_TECHNIQUE;
