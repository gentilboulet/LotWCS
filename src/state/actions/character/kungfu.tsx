import { createAction } from "typesafe-actions";

import * as dataKungFu from "../../../data/kungfu";
import { KUNGFU_TYPE } from "../../../data/kungfu/types";
import { bonusCultivation } from "../../../perks/actions/bonuses";
import { ICost } from "../../models/character/costs";
import { historyMetaCreator } from "../meta";

export const openStyle = createAction(
  "kungfu/OPEN_STYLE",
  (uid: string, kungfuType: KUNGFU_TYPE, cost: ICost) => {
    dataKungFu.validateKungFuType(kungfuType);
    dataKungFu.validateKungFuStyle(kungfuType, uid);

    return {
      cost,
      uid,
      kungfuType,
      perks: [
        bonusCultivation(
          dataKungFu.getKungfuChi(kungfuType, uid),
          cost.original,
        ),
      ],
    };
  },
  historyMetaCreator,
)();

export const buyTechnique = createAction(
  "kungfu/BUY_TECHNIQUE",
  (styleUid: string, uid: string, kungfuType: KUNGFU_TYPE, cost: ICost) => {
    dataKungFu.validateKungFuType(kungfuType);
    dataKungFu.validateKungFuStyle(kungfuType, styleUid);
    dataKungFu.validateKungFuTechnique(kungfuType, styleUid, uid);
    return {
      cost,
      kungfuType,
      styleUid,
      uid,
      perks: [
        bonusCultivation(
          dataKungFu.getKungfuChi(kungfuType, styleUid),
          cost.original,
        ),
      ],
    };
  },
  historyMetaCreator,
)();

export const customStyleName = createAction(
  "kungfu/CUSTOM_NAME_FOR_STYLE",
  (uid: string, kungfuType: KUNGFU_TYPE, name: string) => {
    dataKungFu.validateKungFuType(kungfuType);
    dataKungFu.validateKungFuStyle(kungfuType, uid);
    if (name.length < 1) {
      throw new Error("Invalid new name");
    }
    return {
      kungfuType,
      name,
      uid,
    };
  },
  historyMetaCreator,
)();

export const customTechniqueName = createAction(
  "kungfu/CUSTOM_NAME_FOR_TECHNIQUE",
  (styleUid: string, uid: string, kungfuType: KUNGFU_TYPE, name: string) => {
    dataKungFu.validateKungFuType(kungfuType);
    dataKungFu.validateKungFuStyle(kungfuType, styleUid);
    dataKungFu.validateKungFuTechnique(kungfuType, styleUid, uid);
    if (name.length < 1) {
      throw new Error("Invalid new name");
    }
    return {
      kungfuType,
      name,
      styleUid,
      uid,
    };
  },
  historyMetaCreator,
)();
