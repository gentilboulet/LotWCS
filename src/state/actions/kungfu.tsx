import * as dataKungFu from 'data/kungfu';
import * as constants from 'state/constants/kungfu';
import { ICost } from 'state/costs';

export interface IKungFuOpenStyle {
  type: constants.KUNGFU_OPEN_STYLE;
  cost: ICost;
  uid: string;
  kungfuType: dataKungFu.KUNGFU_TYPE;
}

export interface IKungFuBuyTechnique {
  type: constants.KUNGFU_BUY_TECHNIQUE;
  cost: ICost;
  styleUid: string;
  uid: string;
  kungfuType: dataKungFu.KUNGFU_TYPE;
}

export interface IKungFuSetCustomStyleName {
  type: constants.KUNGFU_CUSTOM_NAME_FOR_STYLE;
  uid: string;
  name: string;
  kungfuType: dataKungFu.KUNGFU_TYPE;
}

export interface IKungFuSetCustomTechniqueName {
  type: constants.KUNGFU_CUSTOM_NAME_FOR_TECHNIQUE;
  styleUid: string;
  uid: string;
  name: string;
  kungfuType: dataKungFu.KUNGFU_TYPE;
}

export type IKungFuAction = IKungFuOpenStyle | IKungFuBuyTechnique
| IKungFuSetCustomStyleName | IKungFuSetCustomTechniqueName

export function openStyle(uid: string, kungfuType: dataKungFu.KUNGFU_TYPE, cost: ICost): IKungFuOpenStyle {
  dataKungFu.validateKungFuType(kungfuType);
  dataKungFu.validateKungFuStyle(kungfuType, uid);
  return { type: constants.KUNGFU_OPEN_STYLE, cost, uid, kungfuType }
}

export function buyTechnique(
  styleUid: string, uid: string, kungfuType: dataKungFu.KUNGFU_TYPE, cost: ICost): IKungFuBuyTechnique {
  dataKungFu.validateKungFuType(kungfuType);
  dataKungFu.validateKungFuStyle(kungfuType, styleUid);
  dataKungFu.validateKungFuTechnique(kungfuType, styleUid, uid);
  return { type: constants.KUNGFU_BUY_TECHNIQUE, cost, styleUid, uid, kungfuType };
}

export function customStyleName(uid: string, kungfuType: dataKungFu.KUNGFU_TYPE, name: string): IKungFuSetCustomStyleName {
  dataKungFu.validateKungFuType(kungfuType);
  dataKungFu.validateKungFuStyle(kungfuType, uid);
  if (name.length < 1) { throw new Error ('Invalid new name'); };
  return { type: constants.KUNGFU_CUSTOM_NAME_FOR_STYLE, uid, name , kungfuType }
}

export function customTechniqueName(
  styleUid: string, uid: string, kungfuType: dataKungFu.KUNGFU_TYPE, name: string
): IKungFuSetCustomTechniqueName {
  dataKungFu.validateKungFuType(kungfuType);
  dataKungFu.validateKungFuStyle(kungfuType, styleUid);
  dataKungFu.validateKungFuTechnique(kungfuType, styleUid, uid);
  if (name.length < 1) { throw new Error ('Invalid new name'); };
  return { type: constants.KUNGFU_CUSTOM_NAME_FOR_TECHNIQUE, styleUid, uid, name , kungfuType }
}
