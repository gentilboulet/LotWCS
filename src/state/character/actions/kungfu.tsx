import { createAction } from "typesafe-actions";

import * as dataKungFu from "../../data/kungfu";
import { KUNGFU_TYPE } from "../../data/kungfu/types";
import { ICost } from "../costs";

export const openStyle = createAction(
  "kungfu/OPEN_STYLE",
  action => (uid: string, kungfuType: KUNGFU_TYPE, cost: ICost) => {
    dataKungFu.validateKungFuType(kungfuType);
    dataKungFu.validateKungFuStyle(kungfuType, uid);
    return action({ uid, kungfuType, cost });
  }
);

export const buyTechnique = createAction(
  "kungfu/BUY_TECHNIQUE",
  action => (
    styleUid: string,
    uid: string,
    kungfuType: KUNGFU_TYPE,
    cost: ICost
  ) => {
    dataKungFu.validateKungFuType(kungfuType);
    dataKungFu.validateKungFuStyle(kungfuType, styleUid);
    dataKungFu.validateKungFuTechnique(kungfuType, styleUid, uid);
    return action({
      cost,
      kungfuType,
      styleUid,
      uid
    });
  }
);

export const customStyleName = createAction(
  "kungfu/CUSTOM_NAME_FOR_STYLE",
  action => (uid: string, kungfuType: KUNGFU_TYPE, name: string) => {
    dataKungFu.validateKungFuType(kungfuType);
    dataKungFu.validateKungFuStyle(kungfuType, uid);
    if (name.length < 1) {
      throw new Error("Invalid new name");
    }
    return action({
      kungfuType,
      name,
      uid
    });
  }
);

export const customTechniqueName = createAction(
  "kungfu/CUSTOM_NAME_FOR_TECHNIQUE",
  action => (
    styleUid: string,
    uid: string,
    kungfuType: KUNGFU_TYPE,
    name: string
  ) => {
    dataKungFu.validateKungFuType(kungfuType);
    dataKungFu.validateKungFuStyle(kungfuType, styleUid);
    dataKungFu.validateKungFuTechnique(kungfuType, styleUid, uid);
    if (name.length < 1) {
      throw new Error("Invalid new name");
    }
    return action({
      kungfuType,
      name,
      styleUid,
      uid
    });
  }
);
