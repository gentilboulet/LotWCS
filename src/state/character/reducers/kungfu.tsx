import { produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../actions/kungfu";
import { ICharacterState } from "../models/type";

import { applyCost } from "../models/costs";

import * as dataKungfu from "../../../data/kungfu";
import * as chi from "../models/chi";
import { addKungFuTechnique, openStyle } from "../models/kungfu";

export function kungfuReducer(
  baseState: ICharacterState,
  action: ActionType<typeof actions>
): ICharacterState {
  switch (action.type) {
    case getType(actions.openStyle):
      return produce(baseState, draftState => {
        applyCost(draftState, action.payload.cost);
        openStyle(
          draftState.kungfu,
          action.payload.kungfuType,
          action.payload.uid
        );

        chi.increaseCultivation(
          draftState.chi,
          dataKungfu.getKungfuChi(
            action.payload.kungfuType,
            action.payload.uid
          ),
          action.payload.cost.original
        );
      });
    case getType(actions.buyTechnique):
      return produce(baseState, draftState => {
        applyCost(draftState, action.payload.cost);
        addKungFuTechnique(
          draftState.kungfu,
          action.payload.kungfuType,
          action.payload.styleUid,
          action.payload.uid
        );

        chi.increaseCultivation(
          draftState.chi,
          dataKungfu.getKungfuChi(
            action.payload.kungfuType,
            action.payload.styleUid
          ),

          action.payload.cost.original
        );
      });
    case getType(actions.customStyleName):
    case getType(actions.customTechniqueName):
    default:
  }
  return baseState;
}
