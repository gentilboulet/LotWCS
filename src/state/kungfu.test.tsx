import * as data from "../data/kungfu";
import { KUNGFU_EXTERNAL, KUNGFU_INTERNAL } from "../data/kungfu/types";
import * as kungfu from "./kungfu";

const _createStateWithAllStyles = () => {
  const state = kungfu.createState();

  data.externalKungfu.forEach(style => {
    kungfu.openStyle(state, KUNGFU_EXTERNAL, style.uid);
  });
  data.internalKungfu.forEach(style => {
    kungfu.openStyle(state, KUNGFU_INTERNAL, style.uid);
  });

  return state;
};

describe("Testing createState", () => {
  it("should create an empty state", () => {
    const state = kungfu.createState();
    expect(state).toMatchSnapshot();
    expect(Object.keys(state.KUNGFU_EXTERNAL).length).toBe(0);
    expect(Object.keys(state.KUNGFU_INTERNAL).length).toBe(0);
  });
});

describe("Testing openStyle", () => {
  it("should open an external style", () => {
    const state = kungfu.createState();
    data.externalKungfu.forEach(style => {
      kungfu.openStyle(state, KUNGFU_EXTERNAL, style.uid);
    });
    const keys = Object.keys(state.KUNGFU_EXTERNAL);
    data.externalKungfu.forEach(style => {
      expect(keys.find(k => k === style.uid)).toBeDefined();
    });
  });
  it("should open an internal style", () => {
    const state = kungfu.createState();
    data.internalKungfu.forEach(style => {
      kungfu.openStyle(state, KUNGFU_INTERNAL, style.uid);
    });
    const keys = Object.keys(state.KUNGFU_INTERNAL);
    data.internalKungfu.forEach(style => {
      expect(keys.find(k => k === style.uid)).toBeDefined();
    });
  });
  it("should not add duplicate styles", () => {
    const state = _createStateWithAllStyles();
    data.internalKungfu.forEach(style => {
      expect(
        kungfu.openStyle(state, KUNGFU_INTERNAL, style.uid)
      ).toThrowError();
    });
  });
});

describe("Testing isStylePresent", () => {
  it("should not find anything in an empty state", () => {
    const state = kungfu.createState();
    data.externalKungfu.forEach(style => {
      expect(
        kungfu.isStylePresent(state, KUNGFU_EXTERNAL, style.uid)
      ).toBeFalsy();
    });
    data.internalKungfu.forEach(style => {
      expect(
        kungfu.isStylePresent(state, KUNGFU_INTERNAL, style.uid)
      ).toBeFalsy();
    });
  });

  it("should detect an existing style", () => {
    const state = _createStateWithAllStyles();

    data.externalKungfu.forEach(style => {
      expect(
        kungfu.isStylePresent(state, KUNGFU_EXTERNAL, style.uid)
      ).toBeTruthy();
    });
    data.internalKungfu.forEach(style => {
      expect(
        kungfu.isStylePresent(state, KUNGFU_INTERNAL, style.uid)
      ).toBeTruthy();
    });
  });
});

describe("Testing addKungFuTechnique", () => {
  it("should add techniques for an already opened style", () => {
    const state = _createStateWithAllStyles();
    data.externalKungfu.forEach(style => {
      style.techniques.forEach(tech => {
        kungfu.addKungFuTechnique(state, KUNGFU_EXTERNAL, style.uid, tech.uid);
        const index = state.KUNGFU_EXTERNAL[style.uid].findIndex(
          stateTechUid => {
            return stateTechUid === tech.uid;
          }
        );
        expect(index).toBeGreaterThanOrEqual(0);
      });
    });
  });
  it("should not add techniques for a not opened style", () => {
    const state = kungfu.createState();
    data.externalKungfu.forEach(style => {
      style.techniques.forEach(tech => {
        expect(() =>
          kungfu.addKungFuTechnique(state, KUNGFU_EXTERNAL, style.uid, tech.uid)
        ).toThrowError();
      });
    });
  });
  it("should not add duplicate techniques", () => {
    const state = _createStateWithAllStyles();
    data.externalKungfu.forEach(style => {
      style.techniques.forEach(tech => {
        kungfu.addKungFuTechnique(state, KUNGFU_EXTERNAL, style.uid, tech.uid);
      });
    });
    expect(state).toMatchSnapshot();
    data.externalKungfu.forEach(style => {
      style.techniques.forEach(tech => {
        expect(() =>
          kungfu.addKungFuTechnique(state, KUNGFU_EXTERNAL, style.uid, tech.uid)
        ).toThrowError();
      });
    });
  });
});

describe("Testing isStyleTechniquePresent", () => {
  it("should not find anything in an empty state", () => {
    const state = kungfu.createState();
    data.externalKungfu.forEach(style => {
      style.techniques.forEach(tech => {
        expect(
          kungfu.isStyleTechniquePresent(
            state,
            KUNGFU_EXTERNAL,
            style.uid,
            tech.uid
          )
        ).toBeFalsy();
      });
    });
    data.internalKungfu.forEach(style => {
      style.techniques.forEach(tech => {
        expect(
          kungfu.isStyleTechniquePresent(
            state,
            KUNGFU_INTERNAL,
            style.uid,
            tech.uid
          )
        ).toBeFalsy();
      });
    });
  });
  it("should detect an existing external style technique", () => {
    const state = _createStateWithAllStyles();
    data.externalKungfu.forEach(style => {
      style.techniques.forEach(tech => {
        kungfu.addKungFuTechnique(state, KUNGFU_EXTERNAL, style.uid, tech.uid);
      });
      style.techniques.forEach(tech => {
        expect(
          kungfu.isStyleTechniquePresent(
            state,
            KUNGFU_EXTERNAL,
            style.uid,
            tech.uid
          )
        ).toBeTruthy();
      });
    });
  });
  it("should detect an existing internal style technique", () => {
    const state = _createStateWithAllStyles();
    data.internalKungfu.forEach(style => {
      style.techniques.forEach(tech => {
        kungfu.addKungFuTechnique(state, KUNGFU_INTERNAL, style.uid, tech.uid);
      });
      style.techniques.forEach(tech => {
        expect(
          kungfu.isStyleTechniquePresent(
            state,
            KUNGFU_INTERNAL,
            style.uid,
            tech.uid
          )
        ).toBeTruthy();
      });
    });
  });
});

describe("Testing getExternalKungFuStatistics", () => {
  it("should return data statistics for a non open style");
  it("should return modified statistics for a style with techniques");
});

describe("Testing canOpenKungFu", () => {
  it("should allow for opening style to an empty state");
  it("should disallow an external style already open");
  it("should disallow an internal style already open");
});

describe("Testing canBuyKungFuTechnique", () => {
  it("should disallow buying any techniques for non opened style");
  it("should allow buying any technique for an external style");
  it(
    "should allow buying techniques in increasing order of level for internal style"
  );
  it(
    "should disallow buying internal technique of an already bought level for a given style"
  );
  it(
    "should allow buying any techniques for an external style whose level  restrictions are lifted"
  );
});
