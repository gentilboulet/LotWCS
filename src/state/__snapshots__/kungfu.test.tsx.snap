// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing addKungFuTechnique should not add duplicate techniques 1`] = `
Object {
  "KUNGFU_EXTERNAL": Object {
    "Blossom Harvest": Array [
      "Heart-Fire Temper Skill",
      "Iron Skin",
      "Iron Skin 2",
      "Nine Mountains Great Strength",
      "No Vermin in the World",
    ],
    "Bone-Fed Wolf Fang": Array [
      "Circling Pack",
      "Fighting With Fire",
      "March Towards Hell",
      "Rainfall at Night",
      "Wolves Devour the Lion",
    ],
  },
  "KUNGFU_INTERNAL": Object {
    "Boundless Prosperity Manual": Array [],
  },
}
`;

exports[`Testing createState should create an empty state 1`] = `
Object {
  "KUNGFU_EXTERNAL": Object {},
  "KUNGFU_INTERNAL": Object {},
}
`;
